### The Libraries

This part of the project is where I will be importing all of the libraries I will use throughout the project.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os
import math

### The Data

This part is intended to just upload all of the data that will be used for the project as dataframes.

In [2]:
fi_2014 = pd.read_csv("Financial_Indicators/2014_Financial_Data.csv", index_col=0)
fi_2015 = pd.read_csv("Financial_Indicators/2015_Financial_Data.csv", index_col=0)
fi_2016 = pd.read_csv("Financial_Indicators/2016_Financial_Data.csv", index_col=0)
fi_2017 = pd.read_csv("Financial_Indicators/2017_Financial_Data.csv", index_col=0)
fi_2018 = pd.read_csv("Financial_Indicators/2018_Financial_Data.csv", index_col=0)

To have some initial idea of the structure of the dataframes I will print the "shape" of them.

In [3]:
print("The Data Shape For 2014's Financial Indicators:\nRows:", fi_2014.shape [0], "\nColumns:", fi_2014.shape [1])
print("\n")

print("The Data Shape For 2015's Financial Indicators:\nRows:", fi_2015.shape [0], "\nColumns:", fi_2015.shape [1])
print("\n")

print("The Data Shape For 2016's Financial Indicators:\nRows:", fi_2016.shape [0], "\nColumns:", fi_2016.shape [1])
print("\n")

print("The Data Shape For 2017's Financial Indicators:\nRows:", fi_2017.shape [0], "\nColumns:", fi_2017.shape [1])
print("\n")

print("The Data Shape For 2018's Financial Indicators:\nRows:", fi_2018.shape [0], "\nColumns:", fi_2018.shape [1])
print("\n")

The Data Shape For 2014's Financial Indicators:
Rows: 3808 
Columns: 224


The Data Shape For 2015's Financial Indicators:
Rows: 4120 
Columns: 224


The Data Shape For 2016's Financial Indicators:
Rows: 4797 
Columns: 224


The Data Shape For 2017's Financial Indicators:
Rows: 4960 
Columns: 224


The Data Shape For 2018's Financial Indicators:
Rows: 4392 
Columns: 224




From this quick check it can be noticed that the number of rows varies with the files from each year (3808, 4120, 4797, 4960, and 4392), which is not a problem since it just represents the companies whose data is in the study, in other words the older files have less stock options than the newer files.  But the amount of columns (features) is the same (224), and that is a good thing because it is an inicial indication that the the same features are present in all files from each year.  

## Data Wrangling

This section is where I will be setting up the data so that it can be used to do the study.

Before going any further, I will perform a quick check of the data calling the info method to see the type of data I will be dealing with and how many non-null value each feature has.

To keep the process organized the datasets will be checked one by one.  I will start by checking the 2014's financial indicator dataset.

In [4]:
fi_2014.info(verbose = True, null_counts = True)

<class 'pandas.core.frame.DataFrame'>
Index: 3808 entries, PG to WTT
Data columns (total 224 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   Revenue                                     3764 non-null   float64
 1   Revenue Growth                              3572 non-null   float64
 2   Cost of Revenue                             3734 non-null   float64
 3   Gross Profit                                3756 non-null   float64
 4   R&D Expenses                                3672 non-null   float64
 5   SG&A Expense                                3749 non-null   float64
 6   Operating Expenses                          3745 non-null   float64
 7   Operating Income                            3753 non-null   float64
 8   Interest Expense                            3745 non-null   float64
 9   Earnings before Tax                         3728 non-null   float64
 10  Income Tax Expen

 As it was noted earlier, the total number of rows in the 2014's financial indicator dataset is 3808, but when looking into the non-null values it can be noticed that most of the features columns do not reach the full 3808 total possible amount.  For the purpose of this project, this is not necessarily a problem if the amount of null values is less than 20% of the total possible amount.  It can also be noted that the grand majority is a float data type.  The only non-float features are the Sector and Class features.  The sector feature indicates which industry the column operated in so ofcourse it is an object data type.  The class feature is listed as an integer data type, but this feature is the one that suggests whether we should invest in the stock (1) or not (0).

Now let's find the info for the other datasets

First the 2015 dataset:

In [5]:
fi_2015.info(verbose = True, null_counts = True)

<class 'pandas.core.frame.DataFrame'>
Index: 4120 entries, PG to WTT
Data columns (total 224 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   Revenue                                     4053 non-null   float64
 1   Revenue Growth                              3780 non-null   float64
 2   Cost of Revenue                             3927 non-null   float64
 3   Gross Profit                                4055 non-null   float64
 4   R&D Expenses                                3855 non-null   float64
 5   SG&A Expense                                3950 non-null   float64
 6   Operating Expenses                          3946 non-null   float64
 7   Operating Income                            4078 non-null   float64
 8   Interest Expense                            3945 non-null   float64
 9   Earnings before Tax                         4017 non-null   float64
 10  Income Tax Expen

The 2016 dataset:

In [6]:
fi_2016.info(verbose = True, null_counts = True)

<class 'pandas.core.frame.DataFrame'>
Index: 4797 entries, PG to WTT
Data columns (total 224 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   Revenue                                     4308 non-null   float64
 1   Revenue Growth                              4148 non-null   float64
 2   Cost of Revenue                             4157 non-null   float64
 3   Gross Profit                                4305 non-null   float64
 4   R&D Expenses                                4084 non-null   float64
 5   SG&A Expense                                4184 non-null   float64
 6   Operating Expenses                          4177 non-null   float64
 7   Operating Income                            4332 non-null   float64
 8   Interest Expense                            4177 non-null   float64
 9   Earnings before Tax                         4266 non-null   float64
 10  Income Tax Expen

The 2017 dataset:

In [7]:
fi_2017.info(verbose = True, null_counts = True)

<class 'pandas.core.frame.DataFrame'>
Index: 4960 entries, PG to WTT
Data columns (total 224 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   Revenue                                     4435 non-null   float64
 1   Revenue Growth                              4236 non-null   float64
 2   Cost of Revenue                             4281 non-null   float64
 3   Gross Profit                                4426 non-null   float64
 4   R&D Expenses                                4173 non-null   float64
 5   SG&A Expense                                4299 non-null   float64
 6   Operating Expenses                          4299 non-null   float64
 7   Operating Income                            4456 non-null   float64
 8   Interest Expense                            4283 non-null   float64
 9   Earnings before Tax                         4381 non-null   float64
 10  Income Tax Expen

And the 2018 dataset:

In [8]:
fi_2018.info(verbose = True, null_counts = True)

<class 'pandas.core.frame.DataFrame'>
Index: 4392 entries, CMCSA to ZYME
Data columns (total 224 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   Revenue                                     4346 non-null   float64
 1   Revenue Growth                              4253 non-null   float64
 2   Cost of Revenue                             4207 non-null   float64
 3   Gross Profit                                4328 non-null   float64
 4   R&D Expenses                                4155 non-null   float64
 5   SG&A Expense                                4226 non-null   float64
 6   Operating Expenses                          4208 non-null   float64
 7   Operating Income                            4357 non-null   float64
 8   Interest Expense                            4208 non-null   float64
 9   Earnings before Tax                         4321 non-null   float64
 10  Income Tax E

The good news is that similar observations done for the 2014 dataset can also be done for the 2015, 2016, 2017, and the 2018 datasets.

The next step in this evaluation will be to check which of the columns in the datasets have 20% or more null values. 

Let's start with the 2014 dataset:

In [9]:
missing_2014 = pd.concat([fi_2014.isnull().sum(), round(100 * fi_2014.isnull().mean(), 2)], axis = 1)
missing_2014.columns = ['Null Count', 'Null %']
print("2014 Null Table\n", missing_2014.sort_values(by = 'Null Count'))

2014 Null Table
                          Null Count  Null %
Class                             0    0.00
Sector                            0    0.00
2015 PRICE VAR [%]                0    0.00
EBIT                              4    0.11
Financing Cash Flow              11    0.29
...                             ...     ...
cashFlowCoverageRatios         1596   41.91
cashFlowToDebtRatio            1596   41.91
shortTermCoverageRatios        1956   51.37
cashConversionCycle            3807   99.97
operatingCycle                 3807   99.97

[224 rows x 2 columns]


On a first look at the null amount count and percentage, the first information that pops up is the fact that only 3 out of the 224 columns.  Though that is a small proportion of non-null values, as it was stated earlier, as long as the columns contains less than 20% of null values it will still be considered eligible for the continuation of the project.  In the other end there are 2 features that has a null value count of 3807, which means those features only have one value out of the possible 3808 that contains a non-null value.  Those features will definitely be deemed useless for the continuation of this project.

Now let's check the null count and percentage of the remaining datasets

In [10]:
missing_2015 = pd.concat([fi_2015.isnull().sum(), round(100 * fi_2015.isnull().mean(), 2)], axis = 1)
missing_2015.columns = ['Null Count', 'Null %']
print("2015 Null Table\n", missing_2015.sort_values(by = 'Null Count'))

print("\n")
print("____________________________________________________________________")
print("\n")

missing_2016 = pd.concat([fi_2016.isnull().sum(), round(100 * fi_2016.isnull().mean(), 2)], axis = 1)
missing_2016.columns = ['Null Count', 'Null %']
print("2016 Null Table\n", missing_2016.sort_values(by = 'Null Count'))

print("\n")
print("____________________________________________________________________")
print("\n")

missing_2017 = pd.concat([fi_2017.isnull().sum(), round(100 * fi_2017.isnull().mean(), 2)], axis = 1)
missing_2017.columns = ['Null Count', 'Null %']
print("2017 Null Table\n", missing_2017.sort_values(by = 'Null Count'))

print("\n")
print("____________________________________________________________________")
print("\n")

missing_2018 = pd.concat([fi_2018.isnull().sum(), round(100 * fi_2018.isnull().mean(), 2)], axis = 1)
missing_2018.columns = ['Null Count', 'Null %']
print("2018 Null Table\n", missing_2018.sort_values(by = 'Null Count'))

2015 Null Table
                                             Null Count  Null %
Class                                                0    0.00
Sector                                               0    0.00
2016 PRICE VAR [%]                                   0    0.00
Financing Cash Flow                                 26    0.63
Net cash flow / Change in cash                      36    0.87
...                                                ...     ...
10Y Revenue Growth (per Share)                    1737   42.16
10Y Shareholders Equity Growth (per Share)        1766   42.86
shortTermCoverageRatios                           1925   46.72
cashConversionCycle                               4118   99.95
operatingCycle                                    4118   99.95

[224 rows x 2 columns]


____________________________________________________________________


2016 Null Table
                                             Null Count  Null %
Class                                             

Once again similar observations made about the 2014 dataset can be made about the other datasets.

The goal of the next step is to isolate the columns that has 20% or more of null values.  This columns will be put in a list to later compare with the other datasets and see which, if any, of the columns are the same.  This comparison will be made prior to deleting them from the dataframe because it is necessary to make sure that every dataset contains, not only the right amount of features, but the exact same features in them.

In [11]:
print("The amount of the 2014 financial indicators dataset columns missing 20% or more values is:",(missing_2014['Null %'] >= 20).sum())

The amount of the 2014 financial indicators dataset columns missing 20% or more values is: 52


There are 52 features in the 2014' financial indicator dataframe that contains 20% or more null values.  So now let's make a list of them.

In [12]:
# Creating a list of the columns from the 2014 dataset that has 20% or more missing values

nullcol_2014 = missing_2014[missing_2014['Null %'] >= 20].index

Now that we have a list of the features that have 20% or more null values for the 2014 financial indicator dataset, I will first make a similar list for the 2015 financial indicator dataset basically following the same process I previously did for the 2014 dataset.

Now to get the same information from the remaining datasets.

In [13]:
print("The amount of the 2015 financial indicators dataset columns missing 20% or more values is:",
      (missing_2015['Null %'] >= 20).sum())

print("\n")

print("The amount of the 2016 financial indicators dataset columns missing 20% or more values is:",
      (missing_2016['Null %'] >= 20).sum())

print("\n")

print("The amount of the 2017 financial indicators dataset columns missing 20% or more values is:",
      (missing_2017['Null %'] >= 20).sum())

print("\n")

print("The amount of the 2018 financial indicators dataset columns missing 20% or more values is:",
      (missing_2018['Null %'] >= 20).sum())

The amount of the 2015 financial indicators dataset columns missing 20% or more values is: 40


The amount of the 2016 financial indicators dataset columns missing 20% or more values is: 58


The amount of the 2017 financial indicators dataset columns missing 20% or more values is: 64


The amount of the 2018 financial indicators dataset columns missing 20% or more values is: 29


From what we can see, each different dataset have a different amounts of features that have 20% or more missing values.  That is not necessarily surprising because, as we noticed earlier, each dataset have different amounts of rows, which means they have some different companies in each dataset.

Now I will save all of those columns that have 20% or more missing values from each dataset in individual lists.

In [14]:
nullcol_2015 = missing_2015[missing_2015['Null %'] >= 20].index

nullcol_2016 = missing_2016[missing_2016['Null %'] >= 20].index

nullcol_2017 = missing_2017[missing_2017['Null %'] >= 20].index

nullcol_2018 = missing_2018[missing_2018['Null %'] >= 20].index

Now I will start comparing the missing values and create new lists with the columns that the datasets have in common and a list with the columns that the datasets do not have in common.  This should be a long process because my plan is to compare, for example, the 2014 dataset with the 2015 dataset as well as the reverse of it, the 2015 dataset with the 2014 dataset.

In [30]:
equal_2014_2015 = []

not_equal_2014_2015 = []

for i in nullcol_2014:
    if i in nullcol_2015:
        equal_2014_2015.append(i)
    else:
        not_equal_2014_2015.append(i)        
        
print("The columns missing 20% or more values in both the 2014 and 2015 are:")
print("_____________________________________________________________________\n")

for x in equal_2014_2015:
    print(x)

print("______________________________________________________________________________________________\n")    
print("The amount of columns of missing 20% or more values in both the 2014 and 2015 datasets is:", len(equal_2014_2015))

The columns missing 20% or more values in both the 2014 and 2015 are:
_____________________________________________________________________

Total non-current assets
Total non-current liabilities
Net Debt
Other Assets
priceCashFlowRatio
priceEarningsToGrowthRatio
enterpriseValueMultiple
ebtperEBIT
niperEBT
effectiveTaxRate
returnOnAssets
returnOnCapitalEmployed
nIperEBT
eBTperEBIT
payablesTurnover
operatingCycle
cashConversionCycle
cashFlowToDebtRatio
freeCashFlowOperatingCashFlowRatio
cashFlowCoverageRatios
shortTermCoverageRatios
dividendPayoutRatio
Net Debt to EBITDA
Current ratio
ROIC
Return on Tangible Assets
Working Capital
Payables Turnover
Weighted Average Shares Diluted Growth
10Y Revenue Growth (per Share)
5Y Revenue Growth (per Share)
10Y Operating CF Growth (per Share)
5Y Operating CF Growth (per Share)
10Y Net Income Growth (per Share)
5Y Net Income Growth (per Share)
10Y Shareholders Equity Growth (per Share)
5Y Shareholders Equity Growth (per Share)
3Y Shareholders Equit

In this step we can see that there are 40 features from both the 2014 and the 2015 datasets that are missing 20% or more values.  The 2014 dataset has a total of 52 features that are missing 20% or more values, as we have seen earlier.  That means that 12 such features of the 2014 that actually contains less missing data in the 2015 dataset.  On the other side, the 2015 dataset has a total of 40 features that have 20% or more missing values, which means all of those 40 features have the same characteristics in the 2014 dataset.

Now let's make the same comparison between the 2014 dataset and the 2016 dataset.

In [31]:
equal_2014_2016 = []

not_equal_2014_2016 = []

for i in nullcol_2014:
    if i in nullcol_2016:
        equal_2014_2016.append(i)
    else:
        not_equal_2014_2016.append(i)        
        
print("The columns missing 20% or more values in both the 2014 and 2016 are:")
print("_____________________________________________________________________\n")

for x in equal_2014_2015:
    print(x)

print("______________________________________________________________________________________________\n")    
print("The amount of columns of missing 20% or more values in both the 2014 and 2016 datasets is:", len(equal_2014_2016))

The columns missing 20% or more values in both the 2014 and 2016 are:
_____________________________________________________________________

Total non-current assets
Total non-current liabilities
Net Debt
Other Assets
priceCashFlowRatio
priceEarningsToGrowthRatio
enterpriseValueMultiple
ebtperEBIT
niperEBT
effectiveTaxRate
returnOnAssets
returnOnCapitalEmployed
nIperEBT
eBTperEBIT
payablesTurnover
operatingCycle
cashConversionCycle
cashFlowToDebtRatio
freeCashFlowOperatingCashFlowRatio
cashFlowCoverageRatios
shortTermCoverageRatios
dividendPayoutRatio
Net Debt to EBITDA
Current ratio
ROIC
Return on Tangible Assets
Working Capital
Payables Turnover
Weighted Average Shares Diluted Growth
10Y Revenue Growth (per Share)
5Y Revenue Growth (per Share)
10Y Operating CF Growth (per Share)
5Y Operating CF Growth (per Share)
10Y Net Income Growth (per Share)
5Y Net Income Growth (per Share)
10Y Shareholders Equity Growth (per Share)
5Y Shareholders Equity Growth (per Share)
3Y Shareholders Equit